src/domain/datasets/domain/src/repos/dataset_entry_repository.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use internal_error::InternalError;
+use opendatafabric::{AccountID, DatasetAlias, DatasetID};
+use thiserror::Error;
+
+use crate::DatasetEntry;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait DatasetEntryRepository: Send + Sync {
+    async fn get_dataset_entry(
+        &self,
+        dataset_id: &DatasetID,
+    ) -> Result<DatasetEntry, GetDatasetEntryError>;
+
+    async fn get_dataset_entries_by_owner_id(
+        &self,
+        owner_id: &AccountID,
+    ) -> Result<Vec<DatasetEntry>, GetDatasetEntryError>;
+
+    async fn save_dataset_entry(&self, dataset: &DatasetEntry)
+        -> Result<(), SaveDatasetEntryError>;
+
+    async fn update_dataset_entry_alias(
+        &self,
+        dataset_id: &DatasetID,
+        new_alias: &DatasetAlias,
+    ) -> Result<(), UpdateDatasetEntryAliasError>;
+
+    async fn delete_dataset_entry(
+        &self,
+        dataset_id: &DatasetID,
+    ) -> Result<(), DeleteEntryDatasetError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum GetDatasetEntryError {
+    #[error(transparent)]
+    NotFound(#[from] DatasetEntryNotFoundError),
+
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+#[derive(Error, Debug)]
+pub enum DatasetEntryNotFoundError {
+    #[error("Dataset entry not found by dataset_id: '{0}'")]
+    ByDatasetId(DatasetID),
+
+    #[error("Datasets entry not found by owner_id: '{0}'")]
+    ByOwnerId(AccountID),
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum SaveDatasetEntryError {
+    #[error(transparent)]
+    Duplicate(#[from] SaveDatasetEntryErrorDuplicate),
+
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+#[derive(Error, Debug)]
+#[error("Dataset entry with dataset_id '{dataset_id}' already exists")]
+pub struct SaveDatasetEntryErrorDuplicate {
+    pub dataset_id: DatasetID,
+}
+
+impl SaveDatasetEntryErrorDuplicate {
+    pub fn new(dataset_id: DatasetID) -> Self {
+        Self { dataset_id }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum UpdateDatasetEntryAliasError {
+    #[error(transparent)]
+    NotFound(#[from] DatasetEntryNotFoundError),
+
+    #[error(transparent)]
+    SameAlias(#[from] DatasetEntryAliasSameError),
```

Comment:
**Q:** Domain question: do we need to check this at the interface/type/contract level?

Reply:
> As per my other comment, dataset names are scoped to owner account, so I think this should not speak about aliases, but rather DatasetName collisions.

yes, for sure

> I'm not sure what you mean by "interface/type/contract level". For naming/renaming to be atomic this will need to be handled on data repository level and an informative error will be necessary to give reason to the caller.

Yes, that was my question -- at what level of the interface should this error be triggered. Or even, more precisely, whether this error is needed in general.

But anyway, I got the answer, thanks!

--

next time, I'll phrase such general questions more clearly

Reply:
PS. for the record, the error is left as is